### ISSUE TITLE
x64: consistent SEGV when using LuaJIT GC64 mode

### ISSUE THREAD

**Issue by agentzh** [openresty/lua-resty-core#144]
See

```
t/request.t .. 1/180
#   Failed test 'TEST 14: ngx.req.set_header (single number value) - status code ok'
#   at /home/agentzh/git/lua-resty-core/../test-nginx/lib/Test/Nginx/Socket.pm line 936.
#          got: ''
#     expected: '200'

#   Failed test 'TEST 14: ngx.req.set_header (single number value) - response_body - response is expected (repeated req 0, req 0)'
#   at /home/agentzh/git/lua-resty-core/../test-nginx/lib/Test/Nginx/Socket.pm line 1501.
#          got: ""
#       length: 0
#     expected: "header foo: 500\x{0a}"
#       length: 16
#     strings begin to differ at char 1 (line 1 column 1)
TEST 14: ngx.req.set_header (single number value) - Can't connect to 127.0.0.1:5855: Connection refused
	Retry connecting after 0.63 sec
TEST 14: ngx.req.set_header (single number value) - Can't connect to 127.0.0.1:5855: Connection refused
	Retry connecting after 0.693 sec
TEST 14: ngx.req.set_header (single number value) - Can't connect to 127.0.0.1:5855: Connection refused
	Retry connecting after 0.759 sec
TEST 14: ngx.req.set_header (single number value) - Can't connect to 127.0.0.1:5855: Connection refused
	Retry connecting after 0.828 sec
TEST 14: ngx.req.set_header (single number value) - Can't connect to 127.0.0.1:5855: Connection refused
	Retry connecting after 0.9 sec
...
```

GDB found no C backtraces:

```
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
0x00007f9f846dfece in ?? ()
(gdb) bt full
#0  0x00007f9f846dfece in ?? ()
No symbol table info available.
#1  0x00007f9f5aec1d08 in ?? ()
No symbol table info available.
#2  0x000000000041c3b0 in ?? ()
No symbol table info available.
#3  0x00007ffde7944390 in ?? ()
No symbol table info available.
#4  0x00007f9f5aeab330 in ?? ()
No symbol table info available.
#5  0x00007f9f5aec56c8 in ?? ()
No symbol table info available.
#6  0x00007ffde7944390 in ?? ()
No symbol table info available.
#7  0x000000000041c3b0 in ?? ()
No symbol table info available.
#8  0x0000000000000000 in ?? ()
No symbol table info available.
```

**Comment by agentzh** [comment]
This issue could not be reproduced when running that test case with Valgrind.

**Comment by agentzh** [comment]
Made more progress by obtaining a minimal and standalone example that can consistently trigger core dumps with the official luajit binary:

https://gist.github.com/agentzh/60b7b3ae8db24934132cbf3978ae5525

Will report it to the luajit mailing list soon.

**Comment by agentzh** [comment]
Mike Pall has already fixed it in the official LuaJIT v2.1 branch. No such segfaults any more. Yay!
